accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,252 @@
+# Simplify RID Model
+
+Date: April 2022
+
+[Runtime IDs (RIDs)](https://docs.microsoft.com/dotnet/core/rid-catalog) are one of the most challenging features of the product. They are necessary, impact many product scenarios, and have very obscure and challenging UX. They also equally affect development and runtime and sometimes prevent .NET from running on operating systems that Microsoft doesn't support. This document proposes a plan to improve some aspects of the way RIDs work, particularly for scenarios that we've found to be insufficient for our needs.
+
+## Context
+
+RIDs are (for the most part) modelled as a graph of [target triplet](https://wiki.osdev.org/Target_Triplet) symbols that describe legal combinations of operating system, chip architecture, and C-runtime, including an extensive fallback scheme. This graph is codified in [`runtime.json`](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json), which is describes the RID catalog. It is a massive (database in a) file. That's a lot of data to reason about, hard to edit correctly, and a bear to maintain.
```

Comment:
> target triplet

AFAIK, target triplets have more cons than pros. It really depends on target audience, but there are negative connotations with the concept of `triplet` in native tooling communities. They consider it vague and archaic concept. e.g. for the modern ISA, the compiler toolchain [documentation](https://www.sifive.com/blog/all-aboard-part-1-compiler-args) was written in terms of ABI (`-mabi`), architecture (`-march`) and microarchitecture (`-mtune`). In case of .NET tooling, `--arch` and `--os` are  relatable concepts.

Reply:
`--arch` and `--os` are indeed nicer to work with. They came out of our effort to make it easier to support emulated environments (like Rosetta 2 on macOS). They are however just a convenience wrapper over the RID concept. More importantly, NuGet packages need a currency for describing assets intended for various. The triplets are the simplest scheme we've found for that. We haven't found a better system.

Reply:
I completely understand and it is probably too late to do anything about it. Since RID is inspired by triplets, I just wanted to point out that modern toolchains prefer explicitly specifying each aspect of platform individually over having a triplet scheme to interpret magic strings, which parser/consumer would always need to remember. They also started their journey with triplets but over the years, found it as an unnecessary overhead.

Reply:
That all makes sense when you are using a compiler to produce artifacts that you will use. I fail to see how that works for a package manager that supports a broad computing ecosystem. Those specific flags also seem too low-level. By definition, they are not intended (per my read) to be used outside of the RISC-V ecosystem.

Reply:
> I fail to see how that works for a package manager that supports a broad computing ecosystem.

`rid: linux-x64` versus `arch: x64, os: linux`

Reply:
OK. You literally just mean using attributes. Got it. Yes, that would have been a different initial design.

Package folks would then look like:

AttributeSpecificAssets
  A
    attributes.json
    Foo.dll
  B
    attributes.json
    Foo.dylib

I'm sure it wouldn't look exactly like that, but that's the basic idea. It could work. A nice characteristic of the current design is that the folder names (for A and B) encode the target so the lookup mechanism is easier. Other things (as you've called out) with that design are worse.

Reply:
Yup, the parsing looks like this: https://github.com/dotnet/runtime/blob/b42adaddff9047e32c9d7f4145746dbb579758f2/src/libraries/Microsoft.NETCore.Platforms/src/RID.cs#L59 and we do that in msbuild scripts and C/C++ separately. :)

Reply:
> AFAIK, target triplets have more cons than pros. It really depends on target audience, but there are negative connotations with the concept of triplet in native tooling communities.

I agree that the classic `machine-vendor-operatingsystem` triplets have archaic feel, due to the `vendor` field and to all sorts of cryptic suffixes that tend to be appended to them.

Most RIDs are `operatingsystem-machine` that does not the same feel. Maybe the doc should not lead with explaining RIDs as target triplets to avoid giving readers wrong impression.

Python `Platform Tag` defined in [PEP 0425[(https://peps.python.org/pep-0425/) is much more similar to RIDs than target triplets. The definition of the Python `Platform Tag` does not make any associations with target triplets.

Reply:
Target triplets do indeed have problems and the parsing is annoying. But I think the fundamental problem is the same -- in order to provide binary compatibility between components, we need to have an encoding for whatever's necessary for binary compatiblity on that platform.

For Mac and Windows, OS version and architecture is probably sufficient. For Linux, we'll also need libc type (and version, although it could be implicitly tied to the TFM).

Reply:
Yes, we are in agreement about what we want to do. This is just a discussion about wording - whether it is helpful to describe RID as target triplet or not.

Reply:
Yes, RID is essentially a close-enough hint and we cannot reliably capture the "desired system configuration" with it accurately. As you pointed out in other thread @agocke, that was not the goal RID was designed to achieve. We perform run time introspections and we have abstraction layers in place to adapt to the system we are running on; in (a fex/countable) places where it makes a difference in dotnet/runtime. This is probably the extent to which 3P package, with OS specific native artifacts,  rely on RID.

Reply:
I am happy to remove the linkage with target triplet if that is what folks want.